text/0000-futures.md

```diff
@@ -154,31 +164,316 @@ can be dropped.
+    /// when the LocalWakers would awaken the same task. However, if this function
+    /// returns true, it is guaranteed that the LocalWakers will awaken the same task.
+    ///
+    /// This function is primarily used for optimization purposes.
+    pub fn will_wake(&self, other: &LocalWaker) -> bool {
+        self.waker == other.waker
+    }
+
+    /// Creates a new `LocalWaker` from `RawWaker`.
+    ///
+    /// The method can not check whether `RawWaker` fulfills the required API
+    /// contract to make it usable for `LocalWaker` and is therefore unsafe.
+    pub unsafe fn new_unchecked(waker: RawWaker) -> LocalWaker {
+        LocalWaker {
+            waker: waker,
+        }
+    }
+}
+
+// The implementations of `Clone` and `Drop` follow what is shown in for `Waker`.
+```
+
+`Waker`s must fulfill the following requirements:
+- They must be cloneable
+- If all instances of a `Waker` have been dropped and their associated task had
+  been driven to completion, all resources which had been allocated for the task
+  must have been released.
+- It must be safe to call `wake()` on a `Waker` even if the associated task has
+  already been driven to completion.
+- `Waker::wake()` must wake up an executor even if it is called from an arbitry
+  thread.
+
+An executor which implements `RawWaker` must therefore make sure that all these
+requirements are fulfilled.
+
+Since many the ownership semantics that are required here can easily be met
+through a reference-counted `Waker` implementation, a convienence method for
+defining `Waker`s is provided, which does not require implementing a `RawWaker`
+and the associated vtable manually.
+
+This convience method is based around the `ArcWake` trait. An implementor of
+an executor can define a type which implements the `ArcWake` trait as defined
+below. The `ArcWake` type defines the associated method `into_local_waker`,
+which allows to retrieve a `LocalWaker` instance from an `Arc` of this type.
+The returned instance will guarantee that the `wake()` and `wake_local` methods
+of the type which implements `ArcWake` are called, whenever `wake()` is called
+on a `Waker` or `LocalWaker`.
 
 ```rust
 /// A way of waking up a specific task.
 ///
-/// Any task executor must provide a way of signaling that a task it owns
-/// is ready to be `poll`ed again. Executors do so by providing a wakeup handle
-/// type that implements this trait.
-pub trait Wake: Send + Sync {
+/// By implementing this trait, types that are expected to be wrapped in an `Arc`
+/// can be converted into `LocalWaker` and `Waker` objects.
+/// Those Wakers can be used to signal executors that a task it owns
+/// is ready to be `poll`ed again.
+pub trait ArcWake: Send + Sync {
     /// Indicates that the associated task is ready to make progress and should
     /// be `poll`ed.
     ///
+    /// This function can called from the thread on which the `ArcWake` was created,
```

Comment:
A possible improvement: "This function can be called from an arbitrary thread, including threads which did not create the `ArcWake`."

Otherwise it has a small type-o "... can *be* called ...".

Reply:
done